File: src/main/java/com/btwrobotics/WhatTime/frc/YearlyMethods/Rebuilt/RebuiltHubManager.java
package com.btwrobotics.WhatTime.frc.YearlyMethods.Rebuilt;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import com.btwrobotics.WhatTime.frc.DashboardManagers.NetworkTablesUtil;
import com.btwrobotics.WhatTime.frc.DriverStation.AllianceManager;
import com.btwrobotics.WhatTime.frc.DriverStation.MatchTimeManager;

import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.DriverStation.Alliance;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;

public class RebuiltHubManager {
    public MatchTimeManager matchTimeManager;

    public RebuiltHubManager(MatchTimeManager matchTimeManager) {
        this.matchTimeManager = matchTimeManager;
    }

    private Optional<Alliance> inactiveFirstAlliance;

    /* Times
     * 0 - Auto: Idk if it has a separate timer for auto
     * 1 - Transition Shift: 2:20-210 or 140s-130s
     * 2 - Phase 1: 2:10-1:45 or 130s-105s
     * 3 - Phase 2: 1:45-1:20 or 105s-80s
     * 4 - Phase 3: 1:20-0:55 or 80s-55s
     * 5 - Phase 4: 0:55-0:30 or 55s-30s
     * 6 - Endgame: 0:30-0:00 or 30s-0s
    */
    public List<Double> sectionStartTimes = new ArrayList<>();


    public Optional<Alliance> currentInactiveAlliance;

    /**
     * Updates the alliance that is inactive first after autonomous.
     */
    public void updateInitialInactiveAlliance() {
        // Gets R or B from DriverStation for which hub is inactive first. Runs at the end of auto.
        String allianceString = DriverStation.getGameSpecificMessage();

        if (allianceString.equals("R")) {
            inactiveFirstAlliance = Optional.of(Alliance.Red);
        }
        else if (allianceString.equals("B")) {
            inactiveFirstAlliance = Optional.of(Alliance.Blue);
        }
    }

    /**
     * Adds triggers for commands at each phase of the match.
     */
    public void scheduleAllInactiveHubChanges() {
        // Transition Shift: 140s-130s
        matchTimeManager.scheduleEventAtTime(
            140, 
            enableAllianceCommand(),
            "Transition Shift"
        );

        // Phase 1: 130s-105s
        matchTimeManager.scheduleEventAtTime(
            130, 
            startInactiveManager(),
            "Teleop Phase 1"
        );

        // Phase 2: 105s-80s
        matchTimeManager.scheduleEventAtTime(
            105, 
            toggleAllianceCommand(),
            "Teleop Phase 2"
        );

        // Phase 3: 80s-55s
        matchTimeManager.scheduleEventAtTime(
            80, 
            toggleAllianceCommand(),
            "Teleop Phase 3"
        );

        // Phase 4: 55s-30s
        matchTimeManager.scheduleEventAtTime(
            55, 
            toggleAllianceCommand(),
            "Teleop Phase 4"
        );

        // Endgame: 30s-0s
        matchTimeManager.scheduleEventAtTime(
            30, 
            enableAllianceCommand(),
            "Endgame"
        );
    }

    // Runs at the start of phase 1 to set the first inactive alliance
    private Command startInactiveManager() {
        return Commands.runOnce(
            () -> {
                currentInactiveAlliance = inactiveFirstAlliance;
            }
        );
    }

    // Switches between inactive alliances
    private Command toggleAllianceCommand() {
        return Commands.runOnce(
            () -> {
                AllianceManager.toggleAlliance(currentInactiveAlliance);
            }
        );
    }

    // Sets the current alliance to active
    private Command enableAllianceCommand() {
        return Commands.runOnce(
            () -> {
                currentInactiveAlliance = AllianceManager.getOpposingAlliance();
            }
        );
    }

    // Sets the current alliance to inactive
    private Command disableAllianceCommand() {
        return Commands.runOnce(
            () -> {
                currentInactiveAlliance = AllianceManager.getCurrentAlliance();
            }
        );
    }

    /**
     * Indicates whether the current alliance's hub is active
     * @return {@code true} when hub is active
     */
    public boolean hubIsActive() {
        return !currentInactiveAlliance.equals(AllianceManager.getCurrentAlliance());
    }

    /**
     * Tells which alliance is inactive first
     * @return an {@code Optional<Alliance>} object of .Red or .Blue
     */
    public Optional<Alliance> getInactiveFirstAlliance() {
        return inactiveFirstAlliance;
    }

    public void putPhasesToNetworkTables() {
        NetworkTablesUtil.put("Upcoming Phases", matchTimeManager.pendingTriggerDescriptions);
        NetworkTablesUtil.put("Past Phases", matchTimeManager.completedTriggerDescriptions);
    }

    public void putPhasesToNetworkTables(String tableName) {
        NetworkTablesUtil.put(tableName, "Upcoming Phases", matchTimeManager.pendingTriggerDescriptions);
        NetworkTablesUtil.put(tableName, "Past Phases", matchTimeManager.completedTriggerDescriptions);
    }
}

File: src/whattime_frc/YearlyMethods/Rebuilt/RebuiltHubManager.py
from __future__ import annotations

from typing import Any

from commands2 import Command
from commands2 import cmd as Commands
from wpilib import DriverStation

from ...DriverStation import AllianceManager

Alliance = DriverStation.Alliance


class RebuiltHubManager:
    def __init__(self, matchTimeManager: Any) -> None:
        self.matchTimeManager = matchTimeManager
        self.inactiveFirstAlliance: Alliance | None = None
        self.currentInactiveAlliance: Alliance | None = None

        # Times:
        # 0 - Auto
        # 1 - Transition Shift: 140s-130s
        # 2 - Phase 1: 130s-105s
        # 3 - Phase 2: 105s-80s
        # 4 - Phase 3: 80s-55s
        # 5 - Phase 4: 55s-30s
        # 6 - Endgame: 30s-0s
        self.sectionStartTimes: list[float] = []

    def updateInitialInactiveAlliance(self) -> None:
        allianceString = DriverStation.getGameSpecificMessage()

        if allianceString == "":
            return

        if allianceString == "R":
            self.inactiveFirstAlliance = Alliance.kRed
        elif allianceString == "B":
            self.inactiveFirstAlliance = Alliance.kBlue

    def scheduleAllInactiveHubChanges(self) -> None:
        self.matchTimeManager.scheduleEventAtTime(
            140,
            self.enableAllianceCommand(),
            "Transition Shift",
        )

        self.matchTimeManager.scheduleEventAtTime(
            130,
            self.startInactiveManager(),
            "Teleop Phase 1",
        )

        self.matchTimeManager.scheduleEventAtTime(
            105,
            self.toggleAllianceCommand(),
            "Teleop Phase 2",
        )

        self.matchTimeManager.scheduleEventAtTime(
            80,
            self.toggleAllianceCommand(),
            "Teleop Phase 3",
        )

        self.matchTimeManager.scheduleEventAtTime(
            55,
            self.toggleAllianceCommand(),
            "Teleop Phase 4",
        )

        self.matchTimeManager.scheduleEventAtTime(
            30,
            self.enableAllianceCommand(),
            "Endgame",
        )

    def startInactiveManager(self) -> Command:
        def _run() -> None:
            if self.inactiveFirstAlliance is None:
                self.updateInitialInactiveAlliance()

            self.currentInactiveAlliance = self.inactiveFirstAlliance

        return Commands.runOnce(_run)

    def toggleAllianceCommand(self) -> Command:
        def _run() -> None:
            if self.currentInactiveAlliance is not None:
                self.currentInactiveAlliance = AllianceManager.toggleAlliance(
                    self.currentInactiveAlliance
                )
            else:
                self.currentInactiveAlliance = AllianceManager.getOpposingAlliance()

        return Commands.runOnce(_run)

    def enableAllianceCommand(self) -> Command:
        def _run() -> None:
            opposingAlliance = AllianceManager.getOpposingAlliance()
            if opposingAlliance is not None:
                self.currentInactiveAlliance = opposingAlliance

        return Commands.runOnce(_run)

    def disableAllianceCommand(self) -> Command:
        def _run() -> None:
            currentAlliance = AllianceManager.getCurrentAlliance()
            if currentAlliance is not None:
                self.currentInactiveAlliance = currentAlliance

        return Commands.runOnce(_run)

    def hubIsActive(self) -> bool:
        if self.currentInactiveAlliance is None:
            return False

        currentAlliance = AllianceManager.getCurrentAlliance()
        if currentAlliance is None:
            return False

        return self.currentInactiveAlliance != currentAlliance

    def getInactiveFirstAlliance(self) -> Alliance | None:
        return self.inactiveFirstAlliance

    # def putPhasesToNetworkTables(self, tableName: str | None = None) -> None:
    #     if NetworkTablesUtil is None:
    #         raise RuntimeError(
    #             "NetworkTablesUtil is not available at "
    #             "whattime_frc.DashboardManagers.NetworkTablesUtil."
    #         )

    #     if tableName is None:
    #         NetworkTablesUtil.put(
    #             "Upcoming Phases",
    #             self.matchTimeManager.pendingTriggerDescriptions,
    #         )
    #         NetworkTablesUtil.put(
    #             "Past Phases",
    #             self.matchTimeManager.completedTriggerDescriptions,
    #         )
    #         return

    #     NetworkTablesUtil.put(
    #         tableName,
    #         "Upcoming Phases",
    #         self.matchTimeManager.pendingTriggerDescriptions,
    #     )
    #     NetworkTablesUtil.put(
    #         tableName,
    #         "Past Phases",
    #         self.matchTimeManager.completedTriggerDescriptions,
    #     )
